src/sbvr-api/permissions.ts

```diff
@@ -0,0 +1,1058 @@
+						constrainedAbstractSqlModel.tables[`${resourceName}$bypass`].definition = createBypassDefinition(table.definition)
+					} else {
+						// Otherwise constrain the non-bypass table
+						onceGetter(table, 'definition', () =>
+							// For $filter on eg a DELETE you need read permissions on the sub-resources,
+							// you only need delete permissions on the resource being deleted
+							generateConstrainedAbstractSql(permissionsLookup, GETPermissions, vocabulary, sqlNameToODataName(table.name))
+						)
+					}
+				})
+				deepFreezeExceptDefinition(constrainedAbstractSqlModel)
+				return constrainedAbstractSqlModel
+			}, {
+				primitive: true
+			}
+		)
+)
+const memoizedGetConstrainedModel = (abstractSqlModel: AbstractSqlModel, permissionsLookup: PermissionLookup, vocabulary: string) =>
+	getBoundConstrainedMemoizer(abstractSqlModel)(permissionsLookup, vocabulary)
+
+export const checkPassword = (username: string, password: string, callback?: (err?: Error, permissions?: string[]) => void): Promise<{
+	id: number
+	actor: number
+	username: string
+	permissions: string[]
+}> => {
+	const authApi = sbvrUtils.api.Auth
+	return authApi.get({
+		resource: 'user',
+		passthrough: {
+			req: rootRead
+		},
+		options: {
+			$select: ['id', 'actor', 'password'],
+			$filter: {
+				username
+			}
+		}
+	})
+	.then((result: AnyObject[]) => {
+		if (result.length === 0) {
+			throw new Error('User not found')
+		}
+		const hash = result[0].password
+		const userId = result[0].id
+		const actorId = result[0].actor
+		return sbvrUtils.sbvrTypes.Hashed.compare(password, hash)
+		.then((res) => {
+			if (!res) {
+				throw new Error('Passwords do not match')
+			}
+			return getUserPermissions(userId)
+				.then((permissions) => {
+					return {
+						id: userId,
+						actor: actorId,
+						username,
+						permissions,
+					}
+				})
+		})
+	})
+	.nodeify(callback)
+}
+
+const getPermissions = (permsFilter: PinejsClientCoreFactory.Filter): Promise<string[]> => {
+	const authApi = sbvrUtils.api.Auth
+	return authApi.get({
+		resource: 'permission',
+		passthrough: {
+			req: rootRead
+		},
+		options: {
+			$select: 'name',
+			$filter: permsFilter,
+			// We orderby to increase the hit rate for the `_checkPermissions` memoisation
+			$orderby: {
+				name: 'asc'
+			}
+		}
+	})
+	.map((permission) => permission.name)
+	.tapCatch((err) => {
+		authApi.logger.error('Error loading permissions', err, err.stack)
+	})
+}
+
+export const getUserPermissions = (userId: number, callback?: (err: Error | undefined, permissions: string[]) => void): Promise<string[]> => {
+	if (_.isString(userId)) {
+		userId = _.parseInt(userId)
+	}
+	if (!_.isFinite(userId)) {
+		return Promise.reject(new Error('User ID has to be numeric, got: ' + typeof userId))
+	}
+	const permsFilter = {
+		$or: {
+			is_of__user: {
+				$any: {
+					$alias: 'uhp',
+					$expr: {
+						uhp: { user: userId },
+						$or: [{
+							uhp: { expiry_date: null }
+						},	{
```

Comment:
A tab slipped in there.

Reply:
Yeah.. I thought prettier was set up on this repo but apparently not.. I'll add it in another PR and then update this one